diff --git a/crypto/boringssl/boringssl.c b/crypto/boringssl/boringssl.c
index 6daedf10..c58a2064 100644
--- a/crypto/boringssl/boringssl.c
+++ b/crypto/boringssl/boringssl.c
@@ -258,6 +258,7 @@ void ngtcp2_crypto_aead_ctx_free(ngtcp2_crypto_aead_ctx *aead_ctx) {
   if (aead_ctx->native_handle) {
     EVP_AEAD_CTX_free(aead_ctx->native_handle);
   }
+  aead_ctx->native_handle = NULL;
 }
 
 typedef struct ngtcp2_crypto_boringssl_cipher_ctx {
@@ -361,7 +362,12 @@ int ngtcp2_crypto_encrypt(uint8_t *dest, const ngtcp2_crypto_aead *aead,
                           const uint8_t *plaintext, size_t plaintextlen,
                           const uint8_t *nonce, size_t noncelen,
                           const uint8_t *aad, size_t aadlen) {
-  const EVP_AEAD *cipher = aead->native_handle;
+  const EVP_AEAD *cipher;
+  if (!aead || !aead_ctx || !aead->native_handle || !aead_ctx->native_handle) {
+    return -1;
+  }
+
+  cipher = aead->native_handle;
   EVP_AEAD_CTX *actx = aead_ctx->native_handle;
   size_t max_outlen = plaintextlen + EVP_AEAD_max_overhead(cipher);
   size_t outlen;
@@ -379,7 +385,12 @@ int ngtcp2_crypto_decrypt(uint8_t *dest, const ngtcp2_crypto_aead *aead,
                           const uint8_t *ciphertext, size_t ciphertextlen,
                           const uint8_t *nonce, size_t noncelen,
                           const uint8_t *aad, size_t aadlen) {
-  const EVP_AEAD *cipher = aead->native_handle;
+  const EVP_AEAD *cipher;
+  if (!aead || !aead_ctx || !aead->native_handle || !aead_ctx->native_handle) {
+    return -1;
+  }
+
+  cipher = aead->native_handle;
   EVP_AEAD_CTX *actx = aead_ctx->native_handle;
   size_t max_overhead = EVP_AEAD_max_overhead(cipher);
   size_t max_outlen;
@@ -403,8 +414,13 @@ int ngtcp2_crypto_hp_mask(uint8_t *dest, const ngtcp2_crypto_cipher *hp,
                           const ngtcp2_crypto_cipher_ctx *hp_ctx,
                           const uint8_t *sample) {
   static const uint8_t PLAINTEXT[16] = {0};
-  ngtcp2_crypto_boringssl_cipher_ctx *ctx = hp_ctx->native_handle;
+  ngtcp2_crypto_boringssl_cipher_ctx *ctx;
   uint32_t counter;
+  if (!hp_ctx || !hp_ctx->native_handle) {
+    return -1;
+  }
+
+  ctx = hp_ctx->native_handle;
 
   (void)hp;
 
